subscriber/common/sink/kafka.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/uber/aresdb/subscriber/config"
+	"github.com/uber/aresdb/utils"
+	"go.uber.org/zap"
+	"strings"
+	"time"
+)
+
+type KafkaPublisher struct {
+	sarama.SyncProducer
+	client.UpsertBatchBuilder
+
+	ServiceConfig config.ServiceConfig
+	JobConfig     *rules.JobConfig
+	Scope         tally.Scope
+	ClusterName   string
+}
+
+func NewKafkaPublisher(serviceConfig config.ServiceConfig, jobConfig *rules.JobConfig, cluster string,
+	sinkCfg config.SinkConfig, aresControllerClient gateway.ControllerClient) (Sink, error) {
+	if sinkCfg.GetSinkMode() != config.Sink_Kafka {
+		return nil, fmt.Errorf("Failed to NewKafkaPublisher, wrong sinkMode=%d", sinkCfg.GetSinkMode())
+	}
+
+	addresses := strings.Split(sinkCfg.KafkaProducerConfig.Brokers, ",")
+	serviceConfig.Logger.Info("Kafka borkers address", zap.Any("brokers", addresses))
+
+	cfg := sarama.NewConfig()
+	cfg.Producer.RequiredAcks = sarama.WaitForAll // Wait for all in-sync replicas to ack the message
+	if sinkCfg.KafkaProducerConfig.RetryMax > 0 {
+		cfg.Producer.Retry.Max = sinkCfg.KafkaProducerConfig.RetryMax
+	}
+	if sinkCfg.KafkaProducerConfig.TimeoutInSec > 0 {
+		cfg.Producer.Timeout = time.Second * time.Duration(sinkCfg.KafkaProducerConfig.TimeoutInSec)
+	}
+	cfg.Producer.Return.Successes = true
+
+	p, err := sarama.NewSyncProducer(addresses, cfg)
+	if err != nil {
+		return nil, utils.StackError(err, "Unable to initialize Kafka producer")
+	}
+
+	// replace httpSchemaFetcher with gateway client
+	// httpSchemaFetcher := NewHttpSchemaFetcher(httpClient, cfg.Address, metricScope)
+	cachedSchemaHandler := client.NewCachedSchemaHandler(
+		serviceConfig.Logger.Sugar(),
+		serviceConfig.Scope.Tagged(map[string]string{
+			"job":         jobConfig.Name,
+			"aresCluster": cluster,
+		}), aresControllerClient)
+
+	// schema refresh is based on job assignment refresh, so disable at here
+	err = cachedSchemaHandler.Start(0)
+	if err != nil {
+		return nil, err
+	}
+
+	kp := KafkaPublisher{
+		SyncProducer: p,
+		UpsertBatchBuilder: client.NewUpsertBatchBuilderImpl(
+			serviceConfig.Logger.Sugar(),
+			serviceConfig.Scope.Tagged(map[string]string{
+				"job":         jobConfig.Name,
+				"aresCluster": cluster,
+			}),
+			cachedSchemaHandler),
+		ServiceConfig: serviceConfig,
+		JobConfig:     jobConfig,
+		Scope: serviceConfig.Scope.Tagged(map[string]string{
+			"job":         jobConfig.Name,
+			"aresCluster": cluster,
+		}),
+		ClusterName: cluster,
+	}
+
+	return &kp, nil
+}
+
+// Shutdown will clean up resources that needs to be cleaned up
+func (kp *KafkaPublisher) Shutdown() {
+	kp.SyncProducer.Close()
+}
+
+// Save saves a batch of row objects into a destination
+func (kp *KafkaPublisher) Save(destination Destination, rows []client.Row) error {
+	shards := Shard(rows, destination, kp.JobConfig)
+	if shards == nil {
+		// case1: no sharding --  publish rows to random kafka partition
+		kp.Insert(destination.Table, -1, destination.ColumnNames, rows, destination.AresUpdateModes...)
```

Comment:
no need to handle insertion error?

Reply:
all errors are already handled in insert function internally (ie. increase error counter)